src/dkg-verify.1.in
.TH DKG\-VERIFY "1" "January 2020" "@PACKAGE_NAME@ @PACKAGE_VERSION@" "User Commands"

.SH NAME
dkg\-verify \-\- verifies a detached signature of a file according to OpenPGP

.SH SYNOPSIS
.B dkg\-verify
.RI [ options ]
.IR KEYFILE

.SH DESCRIPTION
This program verifies a detached signature read from STDIN according to the
OpenPGP standard. KEYFILE represents the (ASCII\-armored) public key, for which
the verification is performed. If KEYFILE is omitted and by option "\-k" a file
containing a set of keys (keyring) is given, the corresponding public key is
determined by the issuer fingerprint or issuer subpacket from the signature.
Optionally with "\-k" some external revocation keys can be provided by the user.
Currently, DSA/DSS, RSA, ECDSA, and EdDSA are permitted as public\-key
algorithms. The input file for which the signature is checked must be specified
by the mandatory option "\-i". Instead of reading the detached signature from
STDIN, by option "\-s" a file containing this signature can be supplied.
.PP
By "\-f" and "\-t" the user can provide a validity period, i.e., signatures
made before the TIMESPEC of option "\-f" or after the TIMESPEC of option "\-t"
are not valid. The corresponding TIMESPEC must be of the format
"YYYY\-MM\-DD[_HH:MM:SS]". The default values are "2009\-06\-01 00:00:00
(publication date of FIPS 186\-3)" and the current system date, respectively.

.PP
The program returns a non\-zero value on error, e.g., if the detached signature
is invalid or no admissible public key is found in KEYFILE.

.SH OPTIONS
Arguments mandatory for long options are also mandatory for short options.
.TP
\fB\-B\fR, \fB\-\-broken\fR
allow broken hash algorithms (i.e. MD5, SHA1, RMD160)
.TP
\fB\-f\fR \fI\,TIMESPEC\/\fR
signature made before given TIMESPEC is not valid
.TP
\fB\-h\fR, \fB\-\-help\fR
print this list of options
.TP
\fB\-i\fR \fI\,FILENAME\/\fR
verify detached signature on FILENAME
.TP
\fB\-k\fR \fI\,FILENAME\/\fR
use keyring FILENAME containing (external revocation) keys
.TP
\fB\-s\fR \fI\,FILENAME\/\fR
read detached signature from FILENAME instead of STDIN
.TP
\fB\-t\fR \fI\,TIMESPEC\/\fR
signature made after given TIMESPEC is not valid
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number of the program
.TP
\fB\-V\fR, \fB\-\-verbose\fR
turn on verbose output
.TP
\fB\-w\fR, \fB\-\-weak\fR
allow weak or expired keys

.SH "SECURITY"
The current implementation is in experimental state and should NOT be used in
production environments.

.SH "REPORTING BUGS"
Security vulnerabilities or cryptographic complaints should be reported
via an email directly to
<HeikoStamer@gmx.net>.
Reported problems will be published after they've been fixed.
.PP
Please report all other bugs via public bug tracker
<https://savannah.nongnu.org/bugs/?func=additem&group=dkgpg>
.PP
Home page of Distributed Privacy Guard (DKGPG):
<https://savannah.nongnu.org/projects/dkgpg/>

.SH "THANKS"
The author thanks Daniel Kahn Gillmor for his valuable suggestions regarding
useful extensions of this program.

.SH "COPYRIGHT"
.TP
Copyright (C)  2017, 2018, 2019, 2020  Heiko Stamer <HeikoStamer@gmx.net>
.PP
Permission is granted to copy, distribute and/or modify this document under
the terms of the GNU Free Documentation License, Version 1.3 or any later
version published by the Free Software Foundation; with no Invariant Sections,
no Front\-Cover Texts, and no Back\-Cover Texts. A copy of the license is
included in the sources of this package and additionally can be obtained from
Internet <https://www.gnu.org/licenses>.

.SH "SEE ALSO"
.BR dkg\-generate (1),
.BR dkg\-sign (1),
.BR dkg\-revoke (1)



NEWS
Copyright (C)  2017, 2018, 2019 Heiko Stamer <HeikoStamer@gmx.net>

Permission is granted to copy, distribute and/or modify this document under
the terms of the GNU Free Documentation License, Version 1.3 or any later
version published by the Free Software Foundation; with no Invariant Sections,
no Front-Cover Texts, and no Back-Cover Texts. A copy of the license is
included in the sources of this package and additionally can be obtained from
the Internet <https://www.gnu.org/licenses>.

1.2.0

1.1.3   This is a bugfix release that includes only three minor improvements: a
        direct-key signature (0x1f) for the primary key is added by default such
        that restricting key servers (e.g. keys.openpgp.org) can deliver a
        cryptographically checkable key without verification of any included
        user ID or without appended subkey. The command line interface of
        dkg-decrypt has been improved in order to give users an easy access to
        the symmetric-key decryption mode. An additional option ("-5") for
        dkg-sign allows to generate V5 signatures (cf. draft RFC 4880bis).
1.1.2   This release adds a lot of features to some programs: two new options
        ("-K" and "-f") allow dkg-keysign to read the certification key from a
        keyring instead of a single key block file. Moreover, with option "-a"
        an interactive confirmation by the user is required for each signature.
        Passive support of V5 keys (cf. draft RFC 4880bis) has been added for
        all programs, however, dkg-generate still generates V4 keys only,
        because this new feature of the draft is not widely spread. There is
        also a new encryption capability: an empty KEYSPEC tells dkg-encrypt to
        create a symmetric-key encrypted session key, i.e., the user has to
        supply a passphrase for encryption and decryption without any public-key
        cryptography involved. Last but not least, two bugs have been fixed:
        First, dkg-decrypt failed on many ZIP-compressed OpenPGP messages with 
        "ZLIB ERROR: -3 invalid block type" due to a bug in decompression logic.
        Second, dkg-decrypt failed in a special case of symmetric-key encrypted
        session keys. Finally, the non-installing program dkg-fuzzer (generates
        fuzzy samples of somehow corrupted OpenPGP stuctures) has been added.
1.1.1   Some small improvements have been applied for dkg-generate: Two new
        options ("-u" and "-N") allow providing the initial user ID and to
        disable the passphrase at command line. Moreover, since this release
        dkg-timestamp and dkg-timestamp-verify require a special key usage flag
        from recent RFC 4880bis draft to select so-called timestamping keys.
        Finally, the synchronization time of the internally used broadcast
        protocol was reduced to a more reasonable amount and in dkg-decrypt the
        detection of end of data for message and decryption shares was changed.
1.1.0   In this release AEAD support (cf. draft RFC 4880bis) has been added
        for dkg-encrypt and dkg-decrypt. Please note that it requires LibTMCG
        version >= 1.3.16 and at least libgcrypt version >= 1.7.0 for OCB mode.
        Moreover, if dkg-generate is called with the new option "--timestamping"
        then a corresponding key usage flag from draft RFC 4880bis is set. 
1.0.9   This release improves the possibilities of DKGPG further. With the new
        programs dkg-adduid and dkg-revuid an user ID can be added and revoked,
        respectively. The program dkg-revoke now supports a human-readable
        reason for revocation (by option "-R") and dkg-decrypt verifies an
        included signature according to a given keyring (option "-k"). Last
        but not least, by the program dkg-addrevoker an external revocation
        key can be specified.
1.0.8   With this release a lot of improvements and new features are introduced.
        First of all, passive support for ECDSA, ECDH, and EdDSA (cf. RFC 6637
        and Werner Koch's draft RFC 4880bis) has been added by relying on the
        most recent version of LibTMCG. The threshold signature scheme and the
        threshold encryption are still limited to finite field cryptography
        (i.e. DSA and ElGamal). Moreover, the programs generate and recognize
        a few other new OpenPGP features (e.g. issuer fingerprint subpackets)
        from RFC 4880bis. Compressed messages are now decompressed by the
        program dkg-decrypt using zlib Compression Library (and optionally by
        library routines from libbzip2). This completes DKGPG's compatibility
        with other OpenPGP software, however, the prefered compression algorithm
        (i.e. "no compression") in self-signatures of generated keys is kept
        for now. Support for symmetric-key decryption by dkg-decrypt has been
        added too. The program dkg-verify now reads the signature from a file,
        if option "-s" is used. To keep track of later protocol changes, all
        interactive programs include a version identifier in their common ID of
        the reliable broadcast channel. Thus programs from previous releases
        will not communicate with those of this release. With the new programs
        dkg-timestamp and dkg-timestamp-verify a OpenPGP timestamp signature
        can be generated and verified, respectively. Last but not least, by the
        new option "-y" some programs (dkg-generate, dkg-decrypt, dkg-sign,
        dkg-keysign, and dkg-timestamp) will work with regular OpenPGP keys too.
        The README file contains a configuration sample showing how to replace
        classic PGP by DKGPG in the famous mail user agent mutt based on this
        option. Please note that this feature is experimental and semantics
        may be changed later.
1.0.7   This release contains a lot of small improvements, in particular due to

src/dkg-verify.cc
/*******************************************************************************
   This file is part of Distributed Privacy Guard (DKGPG).

 Copyright (C) 2017, 2018, 2019, 2020  Heiko Stamer <HeikoStamer@gmx.net>

   DKGPG is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   DKGPG is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with DKGPG; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
*******************************************************************************/

// include headers
#ifdef HAVE_CONFIG_H
	#include "dkgpg_config.h"
#endif

#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <ctime>

#include <libTMCG.hh>
#include "dkg-io.hh"

int main
	(int argc, char **argv)
{
	static const char *usage = "dkg-verify [OPTIONS] -i INPUTFILE KEYFILE";
	static const char *about = PACKAGE_STRING " " PACKAGE_URL;
	static const char *version = PACKAGE_VERSION " (" PACKAGE_NAME ")";

	std::string	filename, ifilename, kfilename, sfilename;
	int 		opt_verbose = 0;
	bool		opt_weak = false, opt_broken = false;
	std::string	sigfrom_str, sigto_str;
	struct tm	sigfrom_tm = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
	struct tm	sigto_tm = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 , 0};
	time_t		sigfrom = 1243810800, sigto = time(NULL);

	// parse command line arguments
	for (size_t i = 0; i < (size_t)(argc - 1); i++)
	{
		std::string arg = argv[i+1];
		if (arg.find("-i") == 0)
		{
			size_t idx = ++i;
			if ((idx < (size_t)(argc - 1)) && (ifilename.length() == 0))
			{
				ifilename = argv[i+1];
			}
			else
			{
				std::cerr << "ERROR: bad option \"" << arg << "\" found" <<
					std::endl;
				return -1;
			}
			continue;
		}
		else if (arg.find("-f") == 0)
		{
			size_t idx = ++i;
			if ((idx < (size_t)(argc - 1)) && (sigfrom_str.length() == 0))
			{
				sigfrom_str = argv[i+1];
			}
			else
			{
				std::cerr << "ERROR: bad option \"" << arg << "\" found" <<
					std::endl;
				return -1;
			}
			continue;
		}
		else if (arg.find("-t") == 0)
		{
			size_t idx = ++i;
			if ((idx < (size_t)(argc - 1)) && (sigto_str.length() == 0))
			{
				sigto_str = argv[i+1];
			}
			else
			{
				std::cerr << "ERROR: bad option \"" << arg << "\" found" <<
					std::endl;
				return -1;
			}
			continue;
		}
		else if (arg.find("-k") == 0)
		{
			size_t idx = ++i;
			if ((idx < (size_t)(argc - 1)) && (kfilename.length() == 0))
			{
				kfilename = argv[i+1];
			}
			else
			{
				std::cerr << "ERROR: bad option \"" << arg << "\" found" <<
					std::endl;
				return -1;
			}
			continue;
		}
		else if (arg.find("-s") == 0)
		{
			size_t idx = ++i;
			if ((idx < (size_t)(argc - 1)) && (sfilename.length() == 0))
			{
				sfilename = argv[i+1];
			}
			else
			{
				std::cerr << "ERROR: bad option \"" << arg << "\" found" <<
					std::endl;
				return -1;
			}
			continue;
		}
		else if ((arg.find("--") == 0) || (arg.find("-v") == 0) ||
				(arg.find("-h") == 0) || (arg.find("-V") == 0) ||
				(arg.find("-w") == 0) || (arg.find("-B") == 0))
		{
			if ((arg.find("-h") == 0) || (arg.find("--help") == 0))
			{
				std::cout << usage << std::endl;
				std::cout << about << std::endl;
				std::cout << "Arguments mandatory for long options are also" <<
					" mandatory for short options." << std::endl;
				std::cout << "  -B, --broken   allow broken hash algorithms" <<
					" (i.e. MD5, SHA1, RMD160)" << std::endl;
				std::cout << "  -f TIMESPEC    signature made before given" <<
					" TIMESPEC is not valid" << std::endl;
				std::cout << "  -h, --help     print this help" << std::endl;
				std::cout << "  -i FILENAME    verify detached signature on" <<
					" FILENAME" << std::endl;
				std::cout << "  -k FILENAME    use keyring FILENAME" <<
					" containing (external revocation) keys" << std::endl;
				std::cout << "  -s FILENAME    read signature from FILENAME" <<
					" instead of STDIN" << std::endl; 
				std::cout << "  -t TIMESPEC    signature made after given" <<
					" TIMESPEC is not valid" << std::endl;
				std::cout << "  -v, --version  print the version number" <<
					std::endl;
				std::cout << "  -V, --verbose  turn on verbose output" <<
					std::endl;
				std::cout << "  -w, --weak     allow weak or expired keys" <<
					std::endl;
				return 0; // not continue
			}
			if ((arg.find("-v") == 0) || (arg.find("--version") == 0))
			{
				std::cout << "dkg-verify v" << version << std::endl;
				return 0; // not continue
			}
			if ((arg.find("-V") == 0) || (arg.find("--verbose") == 0))
				opt_verbose++; // increase verbosity
			if ((arg.find("-w") == 0) || (arg.find("--weak") == 0))
				opt_weak = true;
			if ((arg.find("-B") == 0) || (arg.find("--broken") == 0))
				opt_broken = true;
			continue;
		}
		else if (arg.find("-") == 0)
		{
			std::cerr << "ERROR: unknown option \"" << arg << "\"" << std::endl;
			return -1;
		}
		filename = arg;
	}
#ifdef DKGPG_TESTSUITE
	filename = "Test1_dkg-pub.asc";
	ifilename = "Test1_output.bin";
	sfilename = "Test1_output.sig";
	opt_verbose = 2;
#else
#ifdef DKGPG_TESTSUITE_Y
	filename = "TestY-pub.asc";
	ifilename = "TestY_output.asc";
	sfilename = "TestY_output.sig";
	opt_verbose = 2;
#endif
#endif
	// check command line arguments
	if ((kfilename.length() == 0) && (filename.length() == 0))
	{
		std::cerr << "ERROR: argument KEYFILE missing; usage: " <<
			usage << std::endl;
		return -1;
	}
	if (ifilename.length() == 0)
	{
		std::cerr << "ERROR: mandatory option \"-i\" missing; usage: " <<
			usage << std::endl;
		return -1;
	}
	if (sigfrom_str.length() > 0)
	{
		strptime(sigfrom_str.c_str(), "%Y-%m-%d_%H:%M:%S", &sigfrom_tm);
		sigfrom = mktime(&sigfrom_tm);
		if (sigfrom == ((time_t) -1))
		{
			perror("ERROR: dkg-verify (mktime)");
			std::cerr << "ERROR: cannot convert TIMESPEC; required format:" <<
				" YYYY-MM-DD_HH:MM:SS" << std::endl;
			return -1;
		}
	}
	if (sigto_str.length() > 0)
	{
		strptime(sigto_str.c_str(), "%Y-%m-%d_%H:%M:%S", &sigto_tm);
		sigto = mktime(&sigto_tm);
		if (sigto == ((time_t) -1))
		{
			perror("ERROR: dkg-verify (mktime)");
			std::cerr << "ERROR: cannot convert TIMESPEC; required format:" <<
				" YYYY-MM-DD_HH:MM:SS" << std::endl;
			return -1;
		}
	}

	// initialize LibTMCG
	if (!init_libTMCG())
	{
		std::cerr << "ERROR: initialization of LibTMCG failed" << std::endl;
		return -1;
	}
	if (opt_verbose)
	{
		std::cerr << "INFO: using LibTMCG version " << version_libTMCG() <<
			std::endl;
	}

	// read the (ASCII-armored) public key from KEYFILE
	std::string armored_pubkey;
	if (filename.length() > 0)
	{
		if (!autodetect_file(filename, TMCG_OPENPGP_ARMOR_PUBLIC_KEY_BLOCK,
			armored_pubkey))
		{
			return -1;
		}
	}

	// read the (ASCII-armored) keyring from file
	std::string armored_pubring;
	if (kfilename.length() > 0)
	{
		if (!autodetect_file(kfilename, TMCG_OPENPGP_ARMOR_PUBLIC_KEY_BLOCK,
			armored_pubring))
		{
			return -1;
		}
	}

	// read the (ASCII-armored) signature from stdin or from file
	std::string armored_signature;
	if (sfilename.length() > 0)
	{
		if (!autodetect_file(sfilename, TMCG_OPENPGP_ARMOR_SIGNATURE,
			armored_signature))
		{
			return -1;
		}
	}
	else
	{
		char c;
		while (std::cin.get(c))
			armored_signature += c;
		std::cin.clear();
	}

	// parse the signature
	TMCG_OpenPGP_Signature *signature = NULL;
	bool parse_ok = CallasDonnerhackeFinneyShawThayerRFC4880::
		SignatureParse(armored_signature, opt_verbose, signature);
	if (parse_ok)
	{
		if ((signature->type != TMCG_OPENPGP_SIGNATURE_BINARY_DOCUMENT) &&
			(signature->type != TMCG_OPENPGP_SIGNATURE_CANONICAL_TEXT_DOCUMENT))
		{
			std::cerr << "ERROR: wrong signature type " <<
				(int)signature->type << " found" << std::endl;
			delete signature;
			return -1;
		}
	}
	else
	{
		std::cerr << "ERROR: cannot parse resp. use the provided signature" <<
			std::endl;
		return -1;
	}
	if (opt_verbose)
		signature->PrintInfo();

	// parse the keyring, the public key block and corresponding signatures
	TMCG_OpenPGP_Keyring *ring = NULL;
	if (kfilename.length() > 0)
	{
		int opt_verbose_ring = opt_verbose;
		if (opt_verbose_ring > 0)
			opt_verbose_ring--;
		parse_ok = CallasDonnerhackeFinneyShawThayerRFC4880::
			PublicKeyringParse(armored_pubring, opt_verbose_ring, ring);
		if (!parse_ok)
		{
			std::cerr << "WARNING: cannot use the given keyring" << std::endl;
			ring = new TMCG_OpenPGP_Keyring(); // create an empty keyring
		}
	}
	else
		ring = new TMCG_OpenPGP_Keyring(); // create an empty keyring
	if (filename.length() == 0)
	{
		if (!get_key_by_signature(ring, signature, opt_verbose, armored_pubkey))
		{
			delete ring;
			delete signature;
			return -1;
		}
	}
	TMCG_OpenPGP_Pubkey *primary = NULL;
	parse_ok = CallasDonnerhackeFinneyShawThayerRFC4880::
		PublicKeyBlockParse(armored_pubkey, opt_verbose, primary);
	if (parse_ok)
	{
		primary->CheckSelfSignatures(ring, opt_verbose);
		if (!primary->valid && !opt_weak)
		{
			std::cerr << "ERROR: primary key is not valid" << std::endl;
			delete primary;
			delete ring;
			delete signature;
			return -1;
		}
		primary->CheckSubkeys(ring, opt_verbose);
		if (!opt_weak)
			primary->Reduce(); // keep only valid subkeys
		if (primary->Weak(opt_verbose) && !opt_weak)
		{
			std::cerr << "ERROR: weak primary key is not allowed" << std::endl;
			delete primary;
			delete ring;
			delete signature;
			return -1;
		}
	}
	else
	{
		std::cerr << "ERROR: cannot use the provided public key" << std::endl;
		delete ring;
		delete signature;
		return -1;
	}

	// select corresponding public key of the issuer from subkeys
	bool subkey_selected = false;
	size_t subkey_idx = 0, keyusage = 0;
	time_t ckeytime = 0, ekeytime = 0;
	for (size_t j = 0; j < primary->subkeys.size(); j++)
	{
		if (((primary->subkeys[j]->AccumulateFlags() & 0x02) == 0x02) ||
		    (!primary->subkeys[j]->AccumulateFlags() &&
			((primary->subkeys[j]->pkalgo == TMCG_OPENPGP_PKALGO_RSA) || 
			(primary->subkeys[j]->pkalgo == TMCG_OPENPGP_PKALGO_RSA_SIGN_ONLY) ||
			(primary->subkeys[j]->pkalgo == TMCG_OPENPGP_PKALGO_DSA) ||
			(primary->subkeys[j]->pkalgo == TMCG_OPENPGP_PKALGO_ECDSA))))
		{
			if (CallasDonnerhackeFinneyShawThayerRFC4880::
				OctetsCompare(signature->issuer, primary->subkeys[j]->id))
			{
				subkey_selected = true;
				subkey_idx = j;
				keyusage = primary->subkeys[j]->AccumulateFlags();
				ckeytime = primary->subkeys[j]->creationtime;
				ekeytime = primary->subkeys[j]->expirationtime;
				break;
			}
		}
	}

	// check the primary key, if no admissible subkey has been selected
	if (!subkey_selected)
	{
		if (((primary->AccumulateFlags() & 0x02) != 0x02) &&
		    (!primary->AccumulateFlags() &&
			(primary->pkalgo != TMCG_OPENPGP_PKALGO_RSA) &&
			(primary->pkalgo != TMCG_OPENPGP_PKALGO_RSA_SIGN_ONLY) &&
			(primary->pkalgo != TMCG_OPENPGP_PKALGO_DSA) &&
			(primary->pkalgo != TMCG_OPENPGP_PKALGO_ECDSA)))
		{
			std::cerr << "ERROR: no admissible public key found" << std::endl;
			delete signature;
			delete primary;
			delete ring;
			return -1;
		}
		if (!CallasDonnerhackeFinneyShawThayerRFC4880::
			OctetsCompare(signature->issuer, primary->id))
		{
			std::cerr << "ERROR: no admissible public key found" << std::endl;
			delete signature;
			delete primary;
			delete ring;
			return -1;
		}
		keyusage = primary->AccumulateFlags();
		ckeytime = primary->creationtime;
		ekeytime = primary->expirationtime;
	}
	else
	{
		if (primary->subkeys[subkey_idx]->Weak(opt_verbose) && !opt_weak)
		{
			std::cerr << "ERROR: weak subkey is not allowed" << std::endl;
			delete signature;
			delete primary;
			delete ring;
			return -1;
		}
	}

	// additional validity checks on key and signature
	time_t current_time = time(NULL);
	// 1. key validity time (signatures made before key creation or
	//    after key expiry are not valid)
	if (signature->creationtime < ckeytime)
	{
		std::cerr << "ERROR: signature was made before key creation" <<
			std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	if (ekeytime && (signature->creationtime > (ckeytime + ekeytime)))
	{
		std::cerr << "ERROR: signature was made after key expiry" << std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	// 2. signature validity time (expired signatures are not valid)
	if (signature->expirationtime &&
		(current_time > (signature->creationtime + signature->expirationtime)))
	{
		std::cerr << "ERROR: signature is expired" << std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	// 3. key usage flags (signatures made by keys not with the "signing"
	//    capability are not valid)
	if (!opt_weak && ((keyusage & 0x02) != 0x02))
	{
		std::cerr << "ERROR: corresponding key was not intented for signing" <<
			std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	// 4. key validity time (expired keys are not valid)
	if (!opt_weak && ekeytime && (current_time > (ckeytime + ekeytime)))
	{
		std::cerr << "ERROR: corresponding key is expired" << std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	// 5. signature time (signatures made before the sigfrom or after the sigto
	//    timespec are not valid)
	if (signature->creationtime < sigfrom)
	{
		std::cerr << "ERROR: signature was made before given TIMESPEC" <<
			std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	if (signature->creationtime > sigto)
	{
		std::cerr << "ERROR: signature was made after given TIMESPEC" <<
			std::endl;
		delete signature;
		delete primary;
		delete ring;
		return -2;
	}
	// 6. hash algorithm (reject broken hash algorithms)
	if ((signature->hashalgo == TMCG_OPENPGP_HASHALGO_MD5) ||
	    (signature->hashalgo == TMCG_OPENPGP_HASHALGO_SHA1) ||
		(signature->hashalgo == TMCG_OPENPGP_HASHALGO_RMD160))
	{
		std::string hashname;
		CallasDonnerhackeFinneyShawThayerRFC4880::
			AlgorithmHashTextName(signature->hashalgo, hashname);
		if (opt_broken)
		{
			std::cerr << "WARNING: broken hash algorithm " << hashname <<
				" used for signature" << std::endl;
		}
		else
		{
			if (opt_verbose)
			{
				std::cerr << "ERROR: broken hash algorithm " << hashname <<
					" used for signature" << std::endl;
			}
			delete signature;
			delete primary;
			delete ring;
			return -3;
		}
	}

	// verify signature cryptographically
	bool verify_ok = false;
	if (subkey_selected)
	{
		verify_ok = signature->Verify(primary->subkeys[subkey_idx]->key,
			ifilename, opt_verbose);
	}
	else
		verify_ok = signature->Verify(primary->key, ifilename, opt_verbose);

	// release signature
	delete signature;

	// release primary key and keyring
	delete primary;
	delete ring;

	if (!verify_ok)
	{
		if (opt_verbose)
		{
			std::cerr << "INFO: Bad signature for input file \"" <<
				ifilename << "\"" << std::endl;
		}
		return -3;
	}
	else
	{
		if (opt_verbose)
		{
			std::cerr << "INFO: Good signature for input file \"" <<
				ifilename << "\"" << std::endl;
		}
	}
	return 0;
}
